src/plugins/matrixops/flatten.cpp

```diff
@@ -0,0 +1,313 @@
+                extract_numeric_value_strict(std::move(arg), name_, codename_),
+                order);
+
+        case node_data_type_unknown:
+            return flatten2d(
+                extract_numeric_value(std::move(arg), name_, codename_), order);
+
+        default:
+            break;
+        }
+
+        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+            "phylanx::execution_tree::primitives::flatten::flatten2d",
+            generate_error_message(
+                "the arange primitive requires for all arguments to "
+                "be numeric data types"));
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    hpx::future<primitive_argument_type> flatten::eval(
+        primitive_arguments_type const& operands,
+        primitive_arguments_type const& args, eval_context ctx) const
+    {
+        if (operands.empty() || operands.size() > 2)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "flatten::eval",
+                generate_error_message("the flatten primitive requires exactly "
+                                       "one or two operands"));
+        }
+
+        for (auto const& i : operands)
+        {
+            if (!valid(i))
+            {
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "flatten::eval",
+                    generate_error_message(
+                        "the flatten primitive requires that the "
+                        "arguments given by the operands array are valid"));
+            }
+        }
+
+        auto this_ = this->shared_from_this();
+        if (operands.size() == 1)
+        {
+            return hpx::dataflow(hpx::launch::sync,
+                hpx::util::unwrapping(
+                    [this_ = std::move(this_)](primitive_argument_type&& arg)
+                        -> primitive_argument_type {
+                        std::string order = "C";
+                        if (order != "C" && order != "F")
+                            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                                "phylanx::execution_tree::primitives::flatten::"
+                                "flatten0d",
+                                this_->generate_error_message(
+                                    "order not understood "
+                                    "the order parameter could only be 'C' or "
+                                    "'F'"));
+
+                        std::size_t a_dims = extract_numeric_value_dimension(
+                            arg, this_->name_, this_->codename_);
+
+                        switch (a_dims)
+                        {
+                        case 0:
+                            return this_->flatten0d(std::move(arg));
+
+                        case 1:
+                            return this_->flatten1d(std::move(arg));
+
+                        case 2:
+                            return this_->flatten2d(std::move(arg), order);
```

Comment:
Minor optimization: you might want to move the `order` argument. No need to create another copy. Alternatively, just pass the single character `'C'` or `'F'` to `flatted2d`.